### ISSUE TITLE
New security rules (other type)

### ISSUE THREAD

**Issue by icy17** [the-tcpdump-group/libpcap#1337]
Similar to issue #1336, there are additional types of security rules that are missing in the documentation. I believe incorporating these rules could help developers learn how to correctly use APIs and reduce the risk of security issues.

I will provide a table describing the missing security rules:

| API | parameter index (begin with 1) | Security rules |
|:----------:|:----------:|:----------:|
| pcap_datalink | 1 | parameter must not be freed/closed before |
| pcap_datalink | 1 | Parameter 1: The `p` parameter should not be a handle obtained from opening unauthorized pcap files. |
| pcap_setfilter | 2 | Parameter 2: The caller MUST NOT set the `bf_insns` member of the `bpf_program` structure to NULL before calling `pcap_setfilter` |
| pcap_compile | 3 | Parameter 3: Do not assign or modify the `filter_exp` array in a way that removes or alters its null-termination |
| pcap_compile | 1 | parameter must not be freed/closed before |
| pcap_freecode | 1 | parameter must not be freed/closed before |
| pcap_dump | 1 | parameter must not be freed/closed before |
| pcap_dump | 3 | Parameter 3: Avoid setting `pkt_header.caplen` to a value greater than the allocated space |
| pcap_freealldevs | 1 | parameter must not be freed/closed before |
| pcap_freealldevs | 1 | parameter must not be used later |

**Comment by guyharris** [comment]
> parameter must not be freed/closed before

That applies to *all* calls that take a `pcap_t *` as an argument; it's *not* unique to the ones listed here.

And the parameter should never bee "freed" in the sense of calling `free()` on it. When the program is finished with it, it must call `pcap_close()`; that's the *only* valid way to free it.

> pcap_datalink ... Parameter 1: The p parameter should not be a handle obtained from opening unauthorized pcap files.

What do you mean by "unauthorized pcap files"?  If you mean "capture files to which the process does not have read permission", an attempt to open, for reading, a file to which the process does not have permission to read will result in an error, meaning that, in the case of opening it with `pcap_open_offline()`, you'll get back a null `pcap_t *`.  Passing a null `pcap_t *` as an argument to *any* libpcap routine is an error, for which see [my first comment in #1336](https://github.com/the-tcpdump-group/libpcap/issues/1336#issuecomment-2226105781).

> pcap_setfilter ... Parameter 2: The caller MUST NOT set the bf_insns member of the bpf_program structure to NULL before calling pcap_setfilter

And they also shouldn't set `bf_len` to a value other than the total number of (classic) BPF instructions in the array pointed to by `bf_insns`.

**Comment by icy17** [comment]
> > parameter must not be freed/closed before
> 
> That applies to _all_ calls that take a `pcap_t *` as an argument; it's _not_ unique to the ones listed here.
> 
> And the parameter should never bee "freed" in the sense of calling `free()` on it. When the program is finished with it, it must call `pcap_close()`; that's the _only_ valid way to free it.
> 
> > pcap_datalink ... Parameter 1: The p parameter should not be a handle obtained from opening unauthorized pcap files.
> 
> What do you mean by "unauthorized pcap files"? If you mean "capture files to which the process does not have read permission", an attempt to open, for reading, a file to which the process does not have permission to read will result in an error, meaning that, in the case of opening it with `pcap_open_offline()`, you'll get back a null `pcap_t *`. Passing a null `pcap_t *` as an argument to _any_ libpcap routine is an error, for which see [my first comment in #1336](https://github.com/the-tcpdump-group/libpcap/issues/1336#issuecomment-2226105781).
> 
> > pcap_setfilter ... Parameter 2: The caller MUST NOT set the bf_insns member of the bpf_program structure to NULL before calling pcap_setfilter
> 
> And they also shouldn't set `bf_len` to a value other than the total number of (classic) BPF instructions in the array pointed to by `bf_insns`.

I agree with your thoughts. Many of the security rules I proposed are based on common sense, and I believe it would be helpful to highlight them in the documentation (maybe in a specific section of the documentation), especially for developers new to the library.
